### PR TITLE
Add missing GTF_EXCEPT when importing string length intrinsic

### LINKED ISSUE #11343: [RyuJIT] Accessing Length on a `null` string does not throw NullReferenceException
The following throws in Debug and does not throw in Release. 
Should throw regardless of optimizations.

```csharp
    class Program
    {
        static void Main(string[] args)
        {
            string s = null;
            int unused1 = s.Length; // <- should throw but gets optimized away
        }
    }
```

Tried on Desktop 32/64 bit and on netcoreapp1.1 with same results - NRE in debug, no exceptions in release.

The IL looks like the following

```
.method private hidebysig static 
	void Main (
		string[] args
	) cil managed 
{
	// Method begins at RVA 0x2050
	// Code size 8 (0x8)
	.maxstack 8
	.entrypoint

	IL_0000: ldnull
	IL_0001: callvirt instance int32 [mscorlib]System.String::get_Length()
	IL_0006: pop
	IL_0007: ret
} // end of method Program::Main
```

Note - the `callvirt` is still there and should throw when invoked on a null.

### PR DESCRIPTION
**Pull request by mikedn**
Fixes #11343

### CODE REVIEW

**Comment by mikedn** [comment]
jit-diff summary:
```
Total bytes of diff: 110 (0.00 % of base)
    diff is a regression.
Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes
Top file regressions by size (bytes):
          73 : Microsoft.CodeAnalysis.CSharp.dasm (0.00 % of base)
          29 : Microsoft.CodeAnalysis.dasm (0.00 % of base)
          13 : System.Private.CoreLib.dasm (0.00 % of base)
           8 : Microsoft.CodeAnalysis.VisualBasic.dasm (0.00 % of base)
           2 : System.Reflection.Metadata.dasm (0.00 % of base)
Top file improvements by size (bytes):
         -15 : System.Private.Uri.dasm (-0.02 % of base)
6 total files with size differences (1 improved, 5 regressed).
Top method regessions by size (bytes):
          19 : Microsoft.CodeAnalysis.dasm - PeWriter:WriteDirectory(ref,ref,int,int,int,int,ref):this
          17 : Microsoft.CodeAnalysis.CSharp.dasm - DocumentationCommentCompiler:TrimmedStringStartsWith(ref,ref):bool
           8 : Microsoft.CodeAnalysis.CSharp.dasm - SyntaxTokenWithValueAndTrivia`1:SetDiagnostics(ref):ref:this
           8 : Microsoft.CodeAnalysis.CSharp.dasm - SyntaxTokenWithValueAndTrivia`1:WithLeadingTrivia(ref):ref:this
           8 : Microsoft.CodeAnalysis.CSharp.dasm - SyntaxTokenWithValueAndTrivia`1:WithTrailingTrivia(ref):ref:this
Top method improvements by size (bytes):
          -7 : System.Private.Uri.dasm - Uri:UnescapeDataString(ref):ref
          -6 : System.Private.Uri.dasm - Uri:Unescape(ref):ref:this
          -4 : System.Private.CoreLib.dasm - EventSource:CreateManifestAndDescriptors(ref,ref,ref,int):ref
          -3 : System.Private.Uri.dasm - Uri:get_DnsSafeHost():ref:this
          -1 : System.Private.CoreLib.dasm - String:Join(ref,ref):ref (4 methods)
41 total methods with size differences (6 improved, 35 regressed).
```

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib

**Comment by BruceForstall** [comment]
@dotnet-bot test Tizen armel Cross Debug Build
@dotnet-bot test Tizen armel Cross Release Build
@dotnet-bot test Ubuntu arm Cross Release Build
@dotnet-bot test Ubuntu16.04 arm Cross Debug Build
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test

**Comment by mikedn** [comment]
Adding `GTF_EXCEPT` is what the importer already does for `ldlen` so this should be fine. Anyway...

I went a bit through the diffs and it seems that they mostly consist of moved around code. Or more like code that's no longer moved around due to the added `GTF_EXCEPT`. Size regressions appear to be caused by change in register allocation - spills, reloads, different register choices that increase instruction size (e.g. `r8` instead of `rcx`).

There are also some unexpected improvements:
```asm
       shl      r8d, 1
       add      r8d, 2
; becomes
       lea      r8d, [2*r8+2]
```
No idea how adding `GTF_EXCEPT` causes that. It seems that it should be the other way around.

A small regression that caught my eye:
```asm
       mov      eax, dword ptr [rsp+28H]
       mov      rcx, qword ptr [(reloc)]
       mov      rcx, gword ptr [rcx]
       add      eax, dword ptr [rcx+8]
; becomes
       mov      rax, qword ptr [(reloc)]
       mov      rax, gword ptr [rax]
       mov      eax, dword ptr [rax+8]
       mov      ecx, dword ptr [rsp+28H]
       add      eax, ecx
```
It would seem that `GTF_EXCEPT` has a bigger impact than it should. I suppose that shouldn't be surprising, it's simply easier to just not touch `GTF_EXCEPT` nodes than to move them around to the extent this doesn't cause changes in observable behavior.

**Comment by pgavlin** [comment]
> It would seem that `GTF_EXCEPT` has a bigger impact than it should. I suppose that shouldn't be surprising, it's simply easier to just not touch `GTF_EXCEPT` nodes than to move them around to the extent this doesn't cause changes in observable behavior.

Because it's a side effect flag, `GTF_EXCEPT` can have an outsized effect if it is set early: if a tree with `GTF_EXCEPT` is embedded in a larger tree, the `GTF_EXCEPT` flag will bubble up through the tree's ancestors. Since we cannot yet reliably remove side-effect flags in all cases (see #5999), the backend may be unnecessarily pessimistic about code motion (including the sort of code motion that is used to form RMW operations). It's also possible that the `GTF_EXCEPT` flag is blocking some transforms in `gtSetEvalOrder` or morph; it's not terribly clear without a dump.

**Comment by RussKeldorph** [comment]
@mikedn Would you mind adding a regression test with the PR?

**Comment by mikedn** [comment]
Yikes, should have added a test from the start. It's there now.

File: tests/src/JIT/Regression/JitBlue/GitHub_11343/GitHub_11343.csproj [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    <CodeAnalysisDependentAssemblyPaths Condition=" '$(VS100COMNTOOLS)' != '' " Include="$(VS100COMNTOOLS)..\IDE\PrivateAssemblies">
+      <Visible>False</Visible>
+    </CodeAnalysisDependentAssemblyPaths>
+  </ItemGroup>
+  <PropertyGroup>
+    <DebugType></DebugType>
+    <Optimize>True</Optimize>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="GitHub_10940.cs" />
```
Comment (BruceForstall) [inline tests/src/JIT/Regression/JitBlue/GitHub_11343/GitHub_11343.csproj]: @mikedn got the filename wrong here.

Comment (mikedn) [inline tests/src/JIT/Regression/JitBlue/GitHub_11343/GitHub_11343.csproj]: Fixed. I thought it uses `$(MSBuildProjectName)` but it's only the assembly name that does that.

**Pull request merged**